### DISCUSSION TITLE
MQTT on ESP32 reconnection issue

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11343] [General]
I'm using MQTT on an ESP32, It serves as a client to a broker running on a Node-Red server. I've found an issue when trying to reconnect the client after the server has gone down, then restarted.

On loss of connection to the server my software first makes sure it's still connected to the network, and re-establishes if necessary. 
Next the it recreates a client object and tries to connect. If the TRY connect fails it retries every 15secs. 

Here's the issue it will reconnect to the MQTT broker if the server returns during tries 1 thru 9. For the 10th try and beyond, even when the server returns it fails to connect; it just keeps trying not throwing any errors, other than it not being connected.  Is there some timeout limiting the number of connection re-tries, or has some data stack overflowed. I've even modified the code to test for network connection, client object creation and client connection each 15 sec interval without success. However, with the server running, if I simply reboot the ESP32 it connects.

Does, anyone have suggestions as to why there seems to be a MQTT client connection TRY limit?

MM

UPDATE: I believe I solved the issue.

In the re-connection loop I printed the OSError codes, thru loop 10 the code was "[Errno 113] ECONNABORTED"; after loop 10 it changed to simply "23". On investigation error 23 refers to too many open sockets. Simply placing a myclient.disconnect() in the loop should correct the problem.

MM

FOLLOW_UP: Unfortunately putting myclient.disconnect() in the loop throws an error. Also, tried myclient.init(); that throws a confusing error saying the attribute "init" doesn't exist, but it is listed as a function for "myclient".
Right now the only solution that works is a machine.reset().

MM

**Comment by peterhinch** [comment]
You might like to look at [mqtt_as](https://github.com/peterhinch/micropython-mqtt/tree/master/mqtt_as) which offers a resilient MQTT solution.

**Reply by mmliam** [reply]
Hi Peter,
The issue I'm trying to address is not actually on the client side; it's when the MQTT server goes down (the PC needs to do an update, an extended power loss, etc.). The network has an easy test "net.is_connected", but I haven't seen anything similar for MQTT.

I see mqtt_as has a simple client connection test; "client.is_connected". Is mqtt_as a full implimentation of MQTT that can simply replace mqttsimple.py; in other words is it a call compatible super-set of mqttsimple.py?

MM

UPDATE:
Peter,
In reviewing your mqtt_as "isconnected" function it appears to try and reconnect immediately. I also check the network connection. I was looking for an MQTT connection test only and would handle the re-connections as required. 

Also. if I'm  understanding your code correctly, the "isconnected" is simply testing a variable set during the during various functions such as "connect" & "init". It's not directly detecting whether the client has lost connection do to the MQTT server going down.

MM

**Reply by peterhinch** [reply]
[mqtt-as](https://github.com/peterhinch/micropython-mqtt/tree/master/mqtt_as) handles all outages: of WiFi, broker and (if the broker is remote) loss of internet connectivity.

It is not compatible with the official libraries as these use synchronous code. Synchronous code is not really compatible with detecting broker outages, because doing so involves sending MQTT Ping packets and checking for a response. This needs to happen as a background task.

**Reply by mmliam** [reply]
Hi Peter,

So, using your mqtt_as, how do you test for a lost MQTT server connection; would that be your "broker_up" function? As I've mentioned, it appears that your "isconnected" function is only tested a variable set during the connect or init routines, by itself it doesn't appear to indicate a loss of server.

Thanks, MM

**Reply by peterhinch** [reply]
The presence of the server is detected as follows. The module measures the time since the last communication was received from the server. If this exceeds a threshold, the server is assumed down - [see code](https://github.com/peterhinch/micropython-mqtt/blob./ebc96be35176d66550ff58449bed36aeeb520f80/mqtt_as/mqtt_as.py#L677). If the time from the last response exceeds a shorter time, an MQTT Ping packet is sent. This causes the server to respond with a `PINGRESP`, preventing the outage detection from timing out.

This is standard MQTT practice, and is also used by the broker to determine when to send the Last Will.

**Reply by mmliam** [reply]
Ok, I understand your approach. Mine is different in that the server does not communicate to the module. The module knows the server is down when it receives an error after trying to send it a message. 

Right now my approach is working quite well, but I will review having the server send essentially an "I'm alive" message to the module. But, as it's been said, 'if it ain't broke, don't fix it'.

Thanks for the clarification,
MM

P.S. BTW your "see code" link is broken.

**Reply by peterhinch** [reply]
I don't know what went wrong there. The code is [here](https://github.com/peterhinch/micropython-mqtt/blob/ebc96be35176d66550ff58449bed36aeeb520f80/mqtt_as/mqtt_as.py#L677).

**Reply by mmliam** [reply]
Good link now, thanks

**Reply by mmliam** [reply]
BTW Peter,

It's my understanding that umqtt.simple only supports qos 0, or 1; not 2.  Does your mqtt_as support qos 2 for publish & subscribe?

**Reply by peterhinch** [reply]
No, 0 and 1 only. In most cases the need for qos==2 can be addressed at application level. The sender assigns a unique message ID with the receiver rejecting dupes.
